src_c/base.c

```diff
@@ -264,13 +258,86 @@ pg_register_quit(PyObject *self, PyObject *value)
     Py_RETURN_NONE;
 }
 
+/* init pygame modules, returns 1 if successful, 0 if fail, with PyErr set*/
+static int
+pg_mod_autoinit(const char *modname)
+{
+    PyObject *module, *funcobj, *temp;
+    int ret = 0;
+
+    module = PyImport_ImportModule(modname);
+    if (!module)
+        return 0;
+
+    funcobj = PyObject_GetAttrString(module, "__PYGAMEinit__");
+
+    /* If we could not load __PYGAMEinit__, load init function */
+    if (!funcobj) {
+        PyErr_Clear();
+        funcobj = PyObject_GetAttrString(module, "init");
+    }
+
+    if (funcobj) {
+        temp = PyObject_CallObject(funcobj, NULL);
+        if (temp) {
+            Py_DECREF(temp);
+            ret = 1;
+        }
+    }
+
+    Py_DECREF(module);
+    Py_XDECREF(funcobj);
+    return ret;
+}
+
+/* try to quit pygame modules, errors silenced */
+static void
+pg_mod_autoquit(const char *modname)
+{
+    PyObject *module, *funcobj;
+
+    module = PyImport_ImportModule(modname);
+    if (!module) {
+        PyErr_Clear();
+        return;
+    }
+
+    funcobj = PyObject_GetAttrString(module, "__PYGAMEquit__");
+
+    /* If we could not load __PYGAMEquit__, load quit function */
+    if (!funcobj)
+        funcobj = PyObject_GetAttrString(module, "quit");
+
+    if (funcobj)
+        Py_XDECREF(PyObject_CallObject(funcobj, NULL));
+
+    /* Silence errors */
+    if (PyErr_Occurred())
+        PyErr_Clear();
+
+    Py_DECREF(module);
+    Py_XDECREF(funcobj);
+}
+
 static PyObject *
-pg_init(PyObject *self, PyObject *args)
+pg_init(PyObject *self)
 {
-    PyObject *allmodules, *moduleslist, *dict, *func, *result, *mod;
-    Py_ssize_t loop, num;
+    int i = 0, success = 0, fail = 0;
 
-    int success = 0, fail = 0;
+    /* Put all the module names we want to init in this array */
+    const char *modnames[] = {
+        IMPPREFIX "display", /* Display first, this also inits event,time */
+        IMPPREFIX "joystick",
+        IMPPREFIX "font",
+        IMPPREFIX "freetype",
+        IMPPREFIX "mixer",
+#if IS_SDLv2
+        /* IMPPREFIX "_sdl2.controller", Is this required? Comment for now*/
```

Comment:
I had one question here. Should we init the controller module, when pygame.init is used?
Sinc _sdl2 is an experimental module, it makes sense that it is not implicitly init’d for now
I could change it if required 
Note that the code before this PR would init everything anyway

Reply:
Nothing in the _sdl2 namespace should be imported/initialized automatically IMHO.

Reply:
Nice! So we can leave that line as a comment for now :)